account_billing/models/account_billing.py

```diff
@@ -0,0 +1,258 @@
+            self._context.get('active_ids', [])).mapped('currency_id')
+        if len(currency_ids) > 1:
+            raise ValidationError(
+                _('Please select invoices with same currency'))
+        return currency_ids or self.env.user.company_id.currency_id
+
+    @api.multi
+    def _compute_invoice_related_count(self):
+        self.invoice_related_count = len(self.billing_line_ids)
+
+    @api.multi
+    def name_get(self):
+        result = []
+        for billing in self:
+            result.append((billing.id, (billing.name or 'Draft')))
+        return result
+
+    @api.multi
+    def validate_billing(self):
+        for rec in self:
+            # keep the number in case of a billing reset to draft
+            if not rec.name:
+                # Use the right sequence to set the name
+                if rec.bill_type == 'out_invoice':
+                    sequence_code = 'account.customer.billing'
+                if rec.bill_type == 'in_invoice':
+                    sequence_code = 'account.supplier.billing'
+                rec.name = self.env['ir.sequence'].with_context(
+                    ir_sequence_date=rec.date).next_by_code(sequence_code)
+            rec.write({'state': 'billed'})
+            rec.message_post(body=_('Billing is billed.'))
+        return True
+
+    @api.multi
+    def action_cancel_draft(self):
+        for rec in self:
+            rec.write({'state': 'draft'})
+            rec.message_post(body=_('Billing is reset to draft'))
+        return True
+
+    @api.multi
+    def action_cancel(self):
+        for rec in self:
+            invoice_paid = rec.invoice_ids.filtered(
+                lambda l: l.state == 'paid')
+            if invoice_paid:
+                raise ValidationError(_('Invoice paid already.'))
+            rec.write({'state': 'cancel'})
+            self.message_post(body=_('Billing %s is cancelled') % rec.name)
+        return True
+
+    @api.multi
+    def invoice_relate_billing_tree_view(self):
+        self.ensure_one()
+        if self.bill_type == 'out_invoice':
+            action = self.env.ref('account.action_invoice_tree1')
+        if self.bill_type == 'in_invoice':
+            action = self.env.ref('account.action_vendor_bill_template')
+        result = action.read()[0]
+        result.update({'domain': [('id', 'in', self.invoice_ids.ids)]})
+        return result
+
+
+class AccountBillingLine(models.Model):
+    _name = 'account.billing.line'
+    _description = 'Billing Line'
+
+    billing_id = fields.Many2one(
+        comodel_name='account.billing',
+    )
+    invoice_id = fields.Many2one(
+        comodel_name='account.invoice',
+        string='Invoice Reference',
+        default=lambda self: self._context.get('active_ids', []),
+    )
+    number = fields.Char()
+    invoice_date = fields.Date()
+    due_date = fields.Date()
+    origin = fields.Char(
+        string='Source Document',
+    )
+    total = fields.Float()
```

Comment:
I think most lines can be related field, except
1) invoice_id, 2) total
Right?